Better date generation handling

ninakokalj opened the pull request:
This pull request modifies the `DateGenerator` class and the `detect_datetime_format` function it uses to detect the format of the original date, ensuring that the anonymized date is formatted in the same way. 

I updated `detect_datetime_format` to handle all 9 supported languages by utilizing `dateparser.parse()` to convert the string representation of the date into a `datetime` object. Then, I used `format_datetime` from `babel.dates` to format the date back into a string.

Additionally, I modified the `DateGenerator` class to accept a `language` argument, which specifies the language of the text to be anonymized. I also updated the class's `generate` method to use `format_datetime` from `babel.dates` for formatting the anonymized date.

---

Closes #2

ninakokalj committed:
Added parsing of all supported languages

ninakokalj committed:
Changed the DateGenerator class to be able to handle all languages

eriknovak commented on anonipy/anonymize/generators/date_generator.py:
```diff
@@ -1,11 +1,16 @@
 from .interface import GeneratorInterface
 from ...definitions import Entity
 from ...constants import DATE_TRANSFORM_VARIANTS
+from ...constants import LANGUAGES
```
Merge the two `from ...constants import ...` together, i.e.:

```python
from ...constants import DATE_TRANSFORM_VARIANTS, LANGUAGES
```

eriknovak commented on anonipy/anonymize/generators/date_generator.py:
```diff
@@ -115,23 +120,32 @@ class DateGenerator(GeneratorInterface):
 
-    def __init__(self, *args, date_format: str = "auto", day_sigma: int = 30, **kwargs):
-        """Initializes he date generator.
+    def __init__(self, *args, date_format: str = "auto", lang: Union[str, LANGUAGES] = "en", day_sigma: int = 30, **kwargs):
```
Reorder the input parameters:

```python
def __init__(self, *args, lang: Union[str, LANGUAGES] = "en", date_format: str = "auto", day_sigma: int = 30, **kwargs):
```

eriknovak commented on anonipy/anonymize/generators/date_generator.py:
```diff
@@ -115,23 +120,32 @@ class DateGenerator(GeneratorInterface):
 
         Args:
             date_format: The date format in which the date should be generated. More on date formats [see here](https://www.contensis.com/help-and-docs/guides/querying-your-content/zenql-search/date-formats).
+            lang: The language of the text.
```
Reorder the Args to reflect the input order.

eriknovak commented on anonipy/anonymize/generators/date_generator.py:
```diff
@@ -115,23 +120,32 @@ class DateGenerator(GeneratorInterface):
         super().__init__(*args, **kwargs)
         self.date_format = date_format
+
+        if isinstance(lang, str) and lang in ["en", "de", "el", "nl", "fr", "it", "sl", "es", "uk"]:
```
Can we form the list of languages automatically? So that when we add new languages, we do not need to modify this code. 

Is it possible to create the list from the supported `LANGUAGES` values?

eriknovak commented on anonipy/anonymize/generators/date_generator.py:
```diff
@@ -115,23 +120,32 @@ class DateGenerator(GeneratorInterface):
+        else:
+            raise Exception(f"Unknown lang value: {lang}")
+        
         self.day_sigma = day_sigma
```
Move this line below the line with the `self.date_format = date_format`.

eriknovak commented on anonipy/utils/datetime.py:
```diff
@@ -83,7 +118,7 @@
+def detect_datetime_format(datetime: str, lang: str) -> Tuple[datetime.datetime, str]:
     """Detects the datetime format.
 
     Args:
```
Missing `lang` argument in documentation.

eriknovak commented on anonipy/utils/datetime.py:
```diff
@@ -94,17 +129,41 @@ def detect_datetime_format(datetime: str) -> Tuple[datetime.datetime, str]:
+    datetime = re.sub(r"\s+", " ", datetime).strip()
+
+    # Remove all occurrences of '1°'
+    datetime = re.sub(r"1°", '1', datetime)
```
Change single quotes to double quotes.

eriknovak commented on anonipy/utils/datetime.py:
```diff
@@ -94,17 +129,41 @@ def detect_datetime_format(datetime: str) -> Tuple[datetime.datetime, str]:
             except ValueError:
                 continue
-
+           
```
Remove whitespaces.

eriknovak commented on anonipy/utils/datetime.py:
```diff
@@ -94,17 +129,41 @@ def detect_datetime_format(datetime: str) -> Tuple[datetime.datetime, str]:
+    datetime = re.sub(r"1°", '1', datetime)
+    
+    # Remove all occurrences of 'року'
+    datetime = re.sub(r"\bроку\b", "", datetime).strip()
```
I would add all of this datetime formatting (between lines 132 and 151) into a separate function to format the datetime string.

eriknovak reviewed (changes requested):
Generally looks good! There are some comments that I think need to be addressed.

Also, the packages `babel` and `dateparser` are not added to the `requirements.txt` file. Please add them. Otherwise, `anonipy` will throw errors when processing dates.

ninakokalj committed:
Added new requirements

ninakokalj committed:
Added a list of supported languages

ninakokalj committed:
Fixed detect_datetime_format function

ninakokalj committed:
Fixed errors

ninakokalj commented:
I've implemented the requested changes. Please let me know if there's anything else you'd like adjusted.

eriknovak commented on anonipy/constants.py:
```diff
@@ -41,6 +41,9 @@ class LANGUAGES:
     UKRAINIAN = ("uk", "Ukrainian")
 
 
+SUPPORTED_LANGUAGES = [value[0] for attr, value in LANGUAGES.__dict__.items() if not attr.startswith("_") and isinstance(value, tuple)]
```
[B] I would suggest to add `supported_languages` as a class method of `LANGUAGES`:

```python
class LANGUAGES:
    ...

    @classmethod
    def supported_languages(self) -> List[str]:
        """Return a list of all supported languages.

        Returns:
            The list of all supported languages.

        """
        return [lang[0] for lang in self.__dict__.values() if isinstance(lang, tuple)]
```

And then change all the appropriate files to use it.

eriknovak commented on anonipy/utils/datetime.py:
```diff
@@ -83,28 +116,62 @@
     Args:
         datetime: The datetime string to detect the format.
-
+        lang: The language of the datetime string.
```
Add empty line below `lang`.

eriknovak commented on anonipy/anonymize/generators/date_generator.py:
```diff
@@ -1,11 +1,15 @@
 from .interface import GeneratorInterface
 from ...definitions import Entity
-from ...constants import DATE_TRANSFORM_VARIANTS
+from ...constants import DATE_TRANSFORM_VARIANTS, LANGUAGES, SUPPORTED_LANGUAGES
```
Given the comment [B], remove `SUPPORTED_LANGUAGES`.

eriknovak commented on anonipy/anonymize/generators/date_generator.py:
```diff
@@ -134,6 +140,13 @@ def __init__(self, *args, date_format: str = "auto", day_sigma: int = 30, **kwar
         self.date_format = date_format
         self.day_sigma = day_sigma
 
+        if isinstance(lang, str) and lang in SUPPORTED_LANGUAGES:
```
Given the comment [B], change the line to:

```python
if isinstance(lang, str) and lang in LANGUAGES.supported_languages():
    ...
```

eriknovak reviewed (changes requested):
The code looks better! One suggestion would tidy up the code (I annotated it as [B]). Once that is done, it should be ready to be merged.

ninakokalj committed:
Fixed date parsing errors

ninakokalj committed:
Renamed anonipy/utils/datetime.py to datetime_format.py

ninakokalj commented:
I have implemented the requested changes and renamed the file from `anonipy/utils/datetime.py` to `anonipy/utils/datetime_format.py`. Since I wasn't sure of a better approach, I deleted the old file and created the new one directly in Git.

eriknovak reviewed (approved):
Looks good! Merging.
